---
title: Genotype diagnostics in DO mice
output: rmarkdown::html_vignette
---

```{r knitr_options, echo=FALSE}
knitr::opts_chunk$set(message=FALSE, warning=FALSE, fig.width=7.2,
                      fig.height=5)
options(width=120)
```


QTL analysis results are dependent upon the fidelity of the data, and
so an important preliminary activity is the careful investigation of
the quality of both the genotypes and phenotypes. Diagnosing genotype
data problems can be particularly tricky in multi-parent populations,
such as Diversity Outbred (DO) mouse populations. In this document, we
illustrate our basic strategies for genotype diagnostics in DO mice.

We'll consider data from [Gatti et al. (2014) G3
4:1623-1633](http://www.g3journal.org/content/4/9/1623.full.pdf),
available at
[GitHub](https://github.com/rqtl/qtl2data/tree/master/DO_Gatti2014),
as well as at the [Mouse Phenome
Database](http://phenome.jax.org/db/q?rtn=projects/projdet&reqprojid=289).
We first load the data, directly from the web.

```{r read_data, eval=FALSE}
library(qtl2geno)
file <- "https://raw.githubusercontent.com/rqtl/qtl2data/master/DO_Gatti2014/do.zip"
do <- read_cross2(file)
```

```{r read_data_really, echo=FALSE, results="hide"}
library(qtl2geno)
rdsfile <- "diag_cache/gatti.rds"
if(file.exists(rdsfile)) {
    do <- readRDS(rdsfile)
} else {
    <<read_data>>
    saveRDS(do, file=rdsfile)
}
```

The data concern `r n_ind(do)` DO mice genotyped on the MUGA array,
with genotypes at `r tot_mar(do)` informative markers. The mice are
from generations 3-5 of the DO, and so this lower-resolution
genotyping array is likely sufficient. Later generations have been
genotyped with the MegaMUGA or GigaMUGA, which have an order of
magnitude more markers.


### Sexes

The first thing we check is the sexes of the mice. One can look at
the proportion of homozygous genotype calls on the X chromosome, but
we find it most informative to look at the SNP array intensities for
SNPs on the X and Y chromosomes.

The detailed SNP array data for these mice are available at the
[Mouse Phenome
Database](http://phenome.jax.org/db/q?rtn=projects/projdet&reqprojid=289),
but the file is nearly 500 MB. We've extracted the overall intensities
for SNPs on the X and Y chromosomes and placed them in CSV files at
[GitHub](https://github.com/rqtl/qtl2data/tree/master/DO_Gatti2014/XYint).
We'll work with these extracted data files.

```{r read_xyint, eval=FALSE}
url <- "https://raw.githubusercontent.com/rqtl/qtl2data/master/DO_Gatti2014/XYint/"
xint <- read.csv(paste0(url, "Gatti2014_Xint.csv"))
yint <- read.csv(paste0(url, "Gatti2014_Yint.csv"))
rownames(xint) <- xint[,1]
rownames(yint) <- yint[,1]
xint <- xint[,-1]
yint <- yint[,-1]
```

```{r read_xyint_really, echo=FALSE, results="hide"}
file <- "diag_cache/xyint.RData"
if(file.exists(file)) {
    load(file)
} else {
    <<read_xyint>>
    save(xint, yint, file=file)
}
```

Let's grab the sexes of the mice from the cross object, `do`.

```{r grab_sex}
sex <- do$covar$Sex
names(sex) <- rownames(do$covar)
```

There's a bit of a problem, that the individual IDs in `xint` and
`yint` are like `JAXW202-3168`, while those in the `do` object are
like `JAXW202.3168`. So we'll replace the hyphens with periods, in the
former.

```{r fix_names}
rownames(xint) <- sub("-", ".", rownames(xint), fixed=TRUE)
rownames(yint) <- sub("-", ".", rownames(yint), fixed=TRUE)
```

Let's also reorder the `sex` vector so the names are aligned, and
double-check that they are all aligned.

```{r reorder_sex}
sex <- sex[rownames(xint)]
all(names(sex) == rownames(xint) & names(sex) == rownames(yint))
```

It turns out that many of the SNPs don't do a great job of
differentiating between males and females. So let's first do a t-test
for sex differences in intensities at each SNP, and use only the SNPs
that show a small P-value.

```{r snp_ttests, eval=FALSE}
xp <- apply(xint, 2, function(a) -log10(t.test(a ~ sex)$p.value))
yp <- apply(yint, 2, function(a) -log10(t.test(a ~ sex)$p.value))
```

```{r snp_ttests_really, echo=FALSE, results="hide"}
file <- "diag_cache/pvals.RData"
if(file.exists(file)) {
    load(file)
} else {
   <<snp_ttests>>
   save(xp, yp, file=file)
}
```

Of the `r ncol(xint)` SNPs on the X chromosome, `r sum(xp > 3)` have
-log~10~ p-value > 3, and
of the `r ncol(yint)` SNPs on the Y chromosome, `r sum(yp > 3)` have
-log~10~ p-value > 3.

Let's calculate the average intensities across those SNPs.

```{r ave_intensity}
ave_xint <- rowMeans(xint[,xp > 3])
ave_yint <- rowMeans(yint[,yp > 3])
```

Let's make a scatterplot of these, colored by sex. I'll use `iplot()`
from [R/qtlcharts](http://kbroman.org/qtlcharts), so that the plot is
interactive, in that when hovering over a point, you'll get a tool tip
with the individual ID.

As we'll see, there are `r sum(ave_xint<0.9 & sex=="F")` female mice
with rather low X chromosome intensities. I'm going to plot the points
in a way that the females will appear on top.

```{r set_size, echo=FALSE}
library(qtlcharts)
setScreenSize()
```

```{r plot_xint_v_yint}
library(qtlcharts)
o <- rev(order(sex))
iplot(ave_xint[o], ave_yint[o], indID=names(ave_xint)[o], group=sex[o],
      chartOpts=list(pointcolor=c("Orchid", "slateblue"),
                     xlab="X chr intensity",
                     ylab="Y chr intensity"))
```

The females are below and to the right (high X intensity, low Y
intensity), while the males are above and to the left (low X
intensity, high Y intensity). But there is one female
(`r names(sex)[sex=="F" & ave_xint < 0.9 & ave_yint > 0.6]`)
who sits in the middle of the cloud of males, and there are three
females
(`r paste(names(sex)[sex=="F" & ave_xint < 0.9 & ave_yint < 0.6], collapse=", ")`)
that are in the bottom-left of the figure (low intensities for
both the X and the Y). The males all look okay: there are
`r sum(sex=="M" & ave_xint > 0.9)` males with average X intensity > 0.9.

I'm inclined to think that `r names(sex)[sex=="F" & ave_xint < 0.9 & ave_yint > 0.6]`
is really male; the other three unusual females could be XO, or they
could be poorly behaved DNA samples.

Let's also look at the frequency of heterozygote genotypes on the X
chromosome. We'll make a scatterplot of the proportion of heterozygote
calls on the X chromosome versus the overall proportion of missing
genotypes.

```{r plot_missing_v_xhet}
phetX <- rowSums(do$geno$X == 2)/rowSums(do$geno$X != 0)
pmis <- n_missing(do, summary="proportion")
iplot(pmis, phetX, indID=ind_ids_geno(do), group=do$covar$Sex,
      chartOpts=list(pointcolor=c("Orchid", "slateblue"),
                     xlab="Proportion missing genotypes",
                     ylab="Proportion het on X chr"))
```

There are a number of males with a high proportion of heterozygous
genotype calls on the X chromosome, but they all have high rates of
missing genotypes and so are likely bad DNA samples. The proportion of
heterozygous calls on the X chromosome in females covers a very broad
range, from about 5% to nearly 75%. While this feature can be useful
for identifying females that were misrecorded as males (shouldn't have
many heterozygous X chromosome genotypes but do), it is not so
informative for identifying males that were misrecorded as females.

Among males with <5% missing genotypes, here are the five with the
highest proportion of heterozygous calls on the X chromosome:

```{r males_high_het}
sort(phetX[do$covar$Sex=="M" & pmis < 0.05], decreasing=TRUE)[1:5]
```

The highest is `r round(max(phetX[do$covar$Sex=="M" & pmis < 0.05], 1)*100)`%,
so it's probably not a problem.

Among females with <5% missing genotypes, here are the six with the
lowest proportion of heterozygous calls on the X chromosome:

```{r females_low_het}
sort(phetX[do$covar$Sex=="F" & pmis < 0.05])[1:6]
```

The lowest five include the four problem samples identified above,
based on array intensities, plus also `JAXW202.2490`, with
heterozygous frequency of `r round(phetX["JAXW202.2490"]*100)`%, but
this might just be real.


### Sample duplicates

We next look for pairs of samples with unusually similar genotype
data, which may be sample duplicates or sample contamination.
We use the function `compare_geno()` in
[R/qtl2geno](https://github.com/rqtl/qtl2geno), which uses the raw SNP
data and simply calculates the proportion of matching SNP genotypes
for each pair of mice.

```{r compare_geno}
cg <- compare_geno(do)
```

The `summary()` function will show the pairs with proportions > 0.9.

```{r summary_compare_geno}
summary(cg)
```

There is one pair that has matching genotypes at all but
`r summary(cg)[1,4]` of the `r summary(cg)[1,5]` markers at which they
were both typed. And actually it's a trio
(`r summary(cg)[1,2]`, `r summary(cg)[2,1]`, `r summary(cg)[1,1]`)
with very similar genotypes. And there seems to be another set of four
mice with very similar genotypes.

Let's look at a histogram of the proportions of matching genotypes
across all pairs. The tick marks below help to show what's going on in
the tail of the distribution, and I've highlighted pairs in red that
have ~80% matching genotypes, and in blue pairs that match at >90% of
SNPs.

```{r hist_compare_geno}
p <- cg[upper.tri(cg)]
par(mar=c(5.1,4.1,0.1,0.1))
hist(p, breaks=201, main="", ylab="Count", las=1,
     xlab="Proportion matching genotypes")
rug(p[p < 0.75], col="#00000030")
rug(p[p >= 0.75 & p < 0.9], col="violetred")
rug(p[p >= 0.90], col="slateblue")
```

As we'll see in the next section, these mice all have appreciable
missing genotype data, and so the issue may not be sample duplicates
but rather poor quality DNAs.


### Sample quality

There are three main checks of sample quality: the proportion of
missing data, the inferred number of crossovers, and the estimated
number of genotyping errors.

```{r plot_prop_missing}
pmis_ind <- sort(n_missing(do, "individual", "proportion"), decreasing=TRUE)
iplot(seq(along=pmis_ind), pmis_ind*100,
      chartOpts=list(xlab="Mice (sorted)", ylab="Percent missing genotypes",
                     ylim=c(0,max(pmis_ind)*101)))
```

There are `r sum(pmis_ind > 0.1)` mice with >10% missing genotype
data. Another `r sum(pmis_ind >= 0.05 & pmis_ind <= 0.1)` mice are
missing between 5 and 10% of genotypes.

Also look at genotype frequencies in the raw SNP data, for individuals
with <5% missing genotypes.

```{r geno_freq}
g <- do.call("cbind", do$geno[1:19])
p <- t(apply(g, 1, function(a) { tab <- table(factor(a, levels=0:3))[-1]; tab/sum(tab) }))
itriplot(p[pmis<0.05,], chartOpts=list(labels=c("A", "H","B")))
```

Let's look at estimates of the total number of crossovers, across the
genome, for each mouse. We first calculate the genotype probabilities,
which can take a bit of time.

```{r calc_genoprob, eval=FALSE}
pr <- calc_genoprob(do, error_prob=0.002, map_function="c-f")
```

```{r calc_genoprob_really, echo=FALSE}
file <- "diag_cache/probs.rds"
if(file.exists(file)) {
    pr <- readRDS(file)
} else {
    pr <- calc_genoprob(do, error_prob=0.002, map_function="c-f", cores=0)
    saveRDS(pr, file=file)
}
```

We next find the most probable genotype at each location. (I prefer
this to Viterbi which seeks the most probable sequence of genotypes
along each chromosome, though that is another possibility; see the
`viterbi()` function.) We then use `count_xo()` to get the estimated
numbers of crossovers.

```{r maxmarg}
v <- maxmarg(pr)
n_xo <- count_xo(v)
```

Let's plot the numbers of crossovers across the genome (obtained by
`rowSums()`) against the proportion of missing data.

```{r plot_n_xo}
iplot(pmis_ind[rownames(n_xo)]*100, rowSums(n_xo),
      chartOpts=list(xlab="Percent missing genotypes",
                     ylab="Number of crossovers"))
```

The only clear outliers, in terms of numbers of crossovers, are the
mice with high rates of missing data; these tend to have low estimated
numbers of crossovers, which is perhaps no surprise.


Calculate error lod scores and combine.

```{r calc_errorlod, eval=FALSE}
errlod <- calc_errorlod(do, pr)
errlod <- do.call("cbind", errlod)
```

```{r calc_errorlod_really, echo=FALSE}
file <- "diag_cache/errorlod.rds"
if(file.exists(file)) {
    errlod <- readRDS(file)
} else {
    errlod <- calc_errorlod(do, pr, cores=0)
    errlod <- do.call("cbind", errlod)
    saveRDS(errlod, file)
}
```

Histogram of the error LOD scores. We really only care about values
>4 or so.

```{r hist_errorlod}
hist(errlod, breaks=200, xlab="Error LOD score", ylab="", yaxt="n", main="")
```

Histogram of numbers of problems by individual. No individual has
more than `r max(rowSums(errlod>4))` error LOD scores >4.

```{r hist_errorlod_byind}
hist(rowSums(errlod>4), breaks=seq(-0.1, 13.1, by=0.2),
     xlab="Number of error LOD > 4", ylab="", yaxt="n", main="")
```

The distribution by marker is quite different. The large error LOD
scores are concentrated on a smaller number of markers. Here's a plot
of percent error LOD scores >4, by marker.

```{r plot_errorlod_bymar}
chr <- rep(names(do$gmap), sapply(do$gmap, length))
colme <- colMeans(errlod>4)*100

iplot(1:ncol(errlod), colme,
      indID=paste0(colnames(errlod), " (chr ", chr, ")"),
      chartOpts=list(ylab="Percent of error LOD > 4",
                     xlab="Marker index"))
```

Of the `r length(colme)` markers, there are `r sum(colme>10)` with
>10% problems, and `r sum(colme>5)` with >5% problems.
But `r round(mean(colme==0)*100)`% of markers have no problems at all.


### Marker quality

Proportion missing.

```{r pmis_marker}
pmis_mar <- n_missing(do, "marker", "proportion")
```
