<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">

<head>

<meta charset="utf-8" />
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="generator" content="pandoc" />

<meta name="viewport" content="width=device-width, initial-scale=1">



<title>R/qtl2 input file format</title>



<style type="text/css">code{white-space: pre;}</style>
<style type="text/css">
div.sourceCode { overflow-x: auto; }
table.sourceCode, tr.sourceCode, td.lineNumbers, td.sourceCode {
  margin: 0; padding: 0; vertical-align: baseline; border: none; }
table.sourceCode { width: 100%; line-height: 100%; }
td.lineNumbers { text-align: right; padding-right: 4px; padding-left: 4px; color: #aaaaaa; border-right: 1px solid #aaaaaa; }
td.sourceCode { padding-left: 5px; }
code > span.kw { color: #007020; font-weight: bold; } /* Keyword */
code > span.dt { color: #902000; } /* DataType */
code > span.dv { color: #40a070; } /* DecVal */
code > span.bn { color: #40a070; } /* BaseN */
code > span.fl { color: #40a070; } /* Float */
code > span.ch { color: #4070a0; } /* Char */
code > span.st { color: #4070a0; } /* String */
code > span.co { color: #60a0b0; font-style: italic; } /* Comment */
code > span.ot { color: #007020; } /* Other */
code > span.al { color: #ff0000; font-weight: bold; } /* Alert */
code > span.fu { color: #06287e; } /* Function */
code > span.er { color: #ff0000; font-weight: bold; } /* Error */
code > span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
code > span.cn { color: #880000; } /* Constant */
code > span.sc { color: #4070a0; } /* SpecialChar */
code > span.vs { color: #4070a0; } /* VerbatimString */
code > span.ss { color: #bb6688; } /* SpecialString */
code > span.im { } /* Import */
code > span.va { color: #19177c; } /* Variable */
code > span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code > span.op { color: #666666; } /* Operator */
code > span.bu { } /* BuiltIn */
code > span.ex { } /* Extension */
code > span.pp { color: #bc7a00; } /* Preprocessor */
code > span.at { color: #7d9029; } /* Attribute */
code > span.do { color: #ba2121; font-style: italic; } /* Documentation */
code > span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code > span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code > span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
</style>



<link href="data:text/css;charset=utf-8,body%20%7B%0Abackground%2Dcolor%3A%20%23fff%3B%0Amargin%3A%201em%20auto%3B%0Amax%2Dwidth%3A%20700px%3B%0Aoverflow%3A%20visible%3B%0Apadding%2Dleft%3A%202em%3B%0Apadding%2Dright%3A%202em%3B%0Afont%2Dfamily%3A%20%22Open%20Sans%22%2C%20%22Helvetica%20Neue%22%2C%20Helvetica%2C%20Arial%2C%20sans%2Dserif%3B%0Afont%2Dsize%3A%2014px%3B%0Aline%2Dheight%3A%201%2E35%3B%0A%7D%0A%23header%20%7B%0Atext%2Dalign%3A%20center%3B%0A%7D%0A%23TOC%20%7B%0Aclear%3A%20both%3B%0Amargin%3A%200%200%2010px%2010px%3B%0Apadding%3A%204px%3B%0Awidth%3A%20400px%3B%0Aborder%3A%201px%20solid%20%23CCCCCC%3B%0Aborder%2Dradius%3A%205px%3B%0Abackground%2Dcolor%3A%20%23f6f6f6%3B%0Afont%2Dsize%3A%2013px%3B%0Aline%2Dheight%3A%201%2E3%3B%0A%7D%0A%23TOC%20%2Etoctitle%20%7B%0Afont%2Dweight%3A%20bold%3B%0Afont%2Dsize%3A%2015px%3B%0Amargin%2Dleft%3A%205px%3B%0A%7D%0A%23TOC%20ul%20%7B%0Apadding%2Dleft%3A%2040px%3B%0Amargin%2Dleft%3A%20%2D1%2E5em%3B%0Amargin%2Dtop%3A%205px%3B%0Amargin%2Dbottom%3A%205px%3B%0A%7D%0A%23TOC%20ul%20ul%20%7B%0Amargin%2Dleft%3A%20%2D2em%3B%0A%7D%0A%23TOC%20li%20%7B%0Aline%2Dheight%3A%2016px%3B%0A%7D%0Atable%20%7B%0Amargin%3A%201em%20auto%3B%0Aborder%2Dwidth%3A%201px%3B%0Aborder%2Dcolor%3A%20%23DDDDDD%3B%0Aborder%2Dstyle%3A%20outset%3B%0Aborder%2Dcollapse%3A%20collapse%3B%0A%7D%0Atable%20th%20%7B%0Aborder%2Dwidth%3A%202px%3B%0Apadding%3A%205px%3B%0Aborder%2Dstyle%3A%20inset%3B%0A%7D%0Atable%20td%20%7B%0Aborder%2Dwidth%3A%201px%3B%0Aborder%2Dstyle%3A%20inset%3B%0Aline%2Dheight%3A%2018px%3B%0Apadding%3A%205px%205px%3B%0A%7D%0Atable%2C%20table%20th%2C%20table%20td%20%7B%0Aborder%2Dleft%2Dstyle%3A%20none%3B%0Aborder%2Dright%2Dstyle%3A%20none%3B%0A%7D%0Atable%20thead%2C%20table%20tr%2Eeven%20%7B%0Abackground%2Dcolor%3A%20%23f7f7f7%3B%0A%7D%0Ap%20%7B%0Amargin%3A%200%2E5em%200%3B%0A%7D%0Ablockquote%20%7B%0Abackground%2Dcolor%3A%20%23f6f6f6%3B%0Apadding%3A%200%2E25em%200%2E75em%3B%0A%7D%0Ahr%20%7B%0Aborder%2Dstyle%3A%20solid%3B%0Aborder%3A%20none%3B%0Aborder%2Dtop%3A%201px%20solid%20%23777%3B%0Amargin%3A%2028px%200%3B%0A%7D%0Adl%20%7B%0Amargin%2Dleft%3A%200%3B%0A%7D%0Adl%20dd%20%7B%0Amargin%2Dbottom%3A%2013px%3B%0Amargin%2Dleft%3A%2013px%3B%0A%7D%0Adl%20dt%20%7B%0Afont%2Dweight%3A%20bold%3B%0A%7D%0Aul%20%7B%0Amargin%2Dtop%3A%200%3B%0A%7D%0Aul%20li%20%7B%0Alist%2Dstyle%3A%20circle%20outside%3B%0A%7D%0Aul%20ul%20%7B%0Amargin%2Dbottom%3A%200%3B%0A%7D%0Apre%2C%20code%20%7B%0Abackground%2Dcolor%3A%20%23f7f7f7%3B%0Aborder%2Dradius%3A%203px%3B%0Acolor%3A%20%23333%3B%0Awhite%2Dspace%3A%20pre%2Dwrap%3B%20%0A%7D%0Apre%20%7B%0Aborder%2Dradius%3A%203px%3B%0Amargin%3A%205px%200px%2010px%200px%3B%0Apadding%3A%2010px%3B%0A%7D%0Apre%3Anot%28%5Bclass%5D%29%20%7B%0Abackground%2Dcolor%3A%20%23f7f7f7%3B%0A%7D%0Acode%20%7B%0Afont%2Dfamily%3A%20Consolas%2C%20Monaco%2C%20%27Courier%20New%27%2C%20monospace%3B%0Afont%2Dsize%3A%2085%25%3B%0A%7D%0Ap%20%3E%20code%2C%20li%20%3E%20code%20%7B%0Apadding%3A%202px%200px%3B%0A%7D%0Adiv%2Efigure%20%7B%0Atext%2Dalign%3A%20center%3B%0A%7D%0Aimg%20%7B%0Abackground%2Dcolor%3A%20%23FFFFFF%3B%0Apadding%3A%202px%3B%0Aborder%3A%201px%20solid%20%23DDDDDD%3B%0Aborder%2Dradius%3A%203px%3B%0Aborder%3A%201px%20solid%20%23CCCCCC%3B%0Amargin%3A%200%205px%3B%0A%7D%0Ah1%20%7B%0Amargin%2Dtop%3A%200%3B%0Afont%2Dsize%3A%2035px%3B%0Aline%2Dheight%3A%2040px%3B%0A%7D%0Ah2%20%7B%0Aborder%2Dbottom%3A%204px%20solid%20%23f7f7f7%3B%0Apadding%2Dtop%3A%2010px%3B%0Apadding%2Dbottom%3A%202px%3B%0Afont%2Dsize%3A%20145%25%3B%0A%7D%0Ah3%20%7B%0Aborder%2Dbottom%3A%202px%20solid%20%23f7f7f7%3B%0Apadding%2Dtop%3A%2010px%3B%0Afont%2Dsize%3A%20120%25%3B%0A%7D%0Ah4%20%7B%0Aborder%2Dbottom%3A%201px%20solid%20%23f7f7f7%3B%0Amargin%2Dleft%3A%208px%3B%0Afont%2Dsize%3A%20105%25%3B%0A%7D%0Ah5%2C%20h6%20%7B%0Aborder%2Dbottom%3A%201px%20solid%20%23ccc%3B%0Afont%2Dsize%3A%20105%25%3B%0A%7D%0Aa%20%7B%0Acolor%3A%20%230033dd%3B%0Atext%2Ddecoration%3A%20none%3B%0A%7D%0Aa%3Ahover%20%7B%0Acolor%3A%20%236666ff%3B%20%7D%0Aa%3Avisited%20%7B%0Acolor%3A%20%23800080%3B%20%7D%0Aa%3Avisited%3Ahover%20%7B%0Acolor%3A%20%23BB00BB%3B%20%7D%0Aa%5Bhref%5E%3D%22http%3A%22%5D%20%7B%0Atext%2Ddecoration%3A%20underline%3B%20%7D%0Aa%5Bhref%5E%3D%22https%3A%22%5D%20%7B%0Atext%2Ddecoration%3A%20underline%3B%20%7D%0A%0Acode%20%3E%20span%2Ekw%20%7B%20color%3A%20%23555%3B%20font%2Dweight%3A%20bold%3B%20%7D%20%0Acode%20%3E%20span%2Edt%20%7B%20color%3A%20%23902000%3B%20%7D%20%0Acode%20%3E%20span%2Edv%20%7B%20color%3A%20%2340a070%3B%20%7D%20%0Acode%20%3E%20span%2Ebn%20%7B%20color%3A%20%23d14%3B%20%7D%20%0Acode%20%3E%20span%2Efl%20%7B%20color%3A%20%23d14%3B%20%7D%20%0Acode%20%3E%20span%2Ech%20%7B%20color%3A%20%23d14%3B%20%7D%20%0Acode%20%3E%20span%2Est%20%7B%20color%3A%20%23d14%3B%20%7D%20%0Acode%20%3E%20span%2Eco%20%7B%20color%3A%20%23888888%3B%20font%2Dstyle%3A%20italic%3B%20%7D%20%0Acode%20%3E%20span%2Eot%20%7B%20color%3A%20%23007020%3B%20%7D%20%0Acode%20%3E%20span%2Eal%20%7B%20color%3A%20%23ff0000%3B%20font%2Dweight%3A%20bold%3B%20%7D%20%0Acode%20%3E%20span%2Efu%20%7B%20color%3A%20%23900%3B%20font%2Dweight%3A%20bold%3B%20%7D%20%20code%20%3E%20span%2Eer%20%7B%20color%3A%20%23a61717%3B%20background%2Dcolor%3A%20%23e3d2d2%3B%20%7D%20%0A" rel="stylesheet" type="text/css" />

</head>

<body>




<h1 class="title toc-ignore">R/qtl2 input file format</h1>



<p><a href="http://kbroman.org/qtl2">R/qtl2</a> (aka qtl2) is a reimplementation of the QTL analysis software <a href="https://rqtl.org">R/qtl</a>, to better handle high-dimensional data and complex cross designs.</p>
<p>The input data file formats for <a href="https://rqtl.org">R/qtl</a> cannot handle complex crosses, and so for R/qtl2, we need to define a new input file format. This document describes the details.</p>
<p>For simple cross types, we can continue to use the file formats for <a href="https://rqtl.org">R/qtl</a>, use <code>qtl::read.cross()</code> to read in the data, and then use a conversion function (<code>qtl2geno::convert2cross2()</code>) to convert the data into the new format.</p>
<p>For more complex crosses, we need to define a new format. I was persuaded by <a href="http://aaronwolen.com/">Aaron Wolen</a>’s idea of a <a href="https://github.com/kbroman/qtl/pull/20">“tidy” format</a> for <a href="https://rqtl.org">R/qtl</a>, with three separate CSV files, one for phenotypes, one for genotypes, and one for the genetic map.</p>
<p>Another important idea is from <a href="http://thebird.nl/">Pjotr Prins</a>’s <a href="https://github.com/qtlHD/qtlHD/blob/master/doc/input/qtab.md">qtab format</a>: the inclusion of metadata, such as genotype encodings, with the primary data. This will simplify the handling of multiple files and will help to avoid mistakes.</p>
<p>And so the basic idea for the new format is to have a separate file for each part of the primary data (genotypes, founder genotypes, genetic map, physical map, phenotypes, covariates, and phenotype covariates), and then a control file (in <a href="http://www.yaml.org">YAML</a> or <a href="http://www.json.org">JSON</a> format) which specifies the names of all of those files, the genotype encodings and missing value codes, and things like the name of the sex column within the covariate data (and the encodings for the sexes) and which chromosome is the X chromosome.</p>
<div id="reading-the-data-files" class="section level2">
<h2>Reading the data files</h2>
<p>Before discussing the boring file specifications, let’s consider briefly how the data are read into R.</p>
<p>A key advantage of the control file scheme is that it greatly simplifies the function for reading in the data. That function, <code>read_cross2()</code>, has a <em>single</em> argument: the name (with path) of the control file. So you can read in data like this:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(qtl2geno)
grav2 &lt;-<span class="st"> </span><span class="kw">read_cross2</span>(<span class="st">&quot;~/my_data/grav2.yaml&quot;</span>)</code></pre></div>
<p>The large number of files is a bit cumbersome, so we’ve made it possible to use a <a href="https://en.wikipedia.org/wiki/Zip_(file_format)">zip file</a> containing all of the data files, and to read that zip file directly. There’s even a function for creating the zip file:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">zip_datafiles</span>(<span class="st">&quot;~/my_data/grav2.yaml&quot;</span>)</code></pre></div>
<p>The <code>zip_datafiles()</code> function will read the control file to identify all of the relevant data files and then zip them up into a file with the same name and location, but with the extension <code>.zip</code> rather than <code>.yaml</code> or <code>.json</code>.</p>
<p>To read the data back in, we use the same <code>read_cross2()</code> function, providing the name (and path) of the zip file rather than the control file.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">grav2 &lt;-<span class="st"> </span><span class="kw">read_cross2</span>(<span class="st">&quot;~/my_data/grav2.zip&quot;</span>)</code></pre></div>
<p>This can even be done with remote files.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">grav2 &lt;-<span class="st"> </span><span class="kw">read_cross2</span>(<span class="st">&quot;http://kbroman.org/qtl2/assets/sampledata/grav2/grav2.zip&quot;</span>)</code></pre></div>
<p>Of course, the other advantage of the zip file is that it is <em>compressed</em> and so smaller than the combined set of CSV files.</p>
</div>
<div id="format-of-the-data-files" class="section level2">
<h2>Format of the data files</h2>
<p>The bulk of the data is in a set of comma-delimited (CSV) files. In addition, a control file (in <a href="http://www.yaml.org">YAML</a> or <a href="http://www.json.org">JSON</a> format), contained in the same directory as the CSV files, specifies the file names and other control parameters (such as genotype and sex encodings). Sample data files are available at <a href="http://kbroman.org/qtl2/pages/sampledata.html">the R/qtl2 website</a>. We’ll discuss the CSV files first.</p>
<div id="csv-files" class="section level3">
<h3>CSV files</h3>
<p>The comma-delimited (CSV) files are each in the form of a simple matrix, with the first column being a set of IDs and the first row being a set of variable names.</p>
<p>Missing value codes will be specified in the control file (as <code>na.strings</code>, with default value <code>&quot;NA&quot;</code>) and will apply across all files, so a missing value code for one file cannot be an allowed value in another file.</p>
<p>The CSV files can include a header with a set of comment lines initiated by a value specified in the control file as <code>comment.char</code> (with default value <code>&quot;#&quot;</code>). The first such line could be a description of the contents of the file. These comment lines can include the expected number of rows and columns, like this:</p>
<pre><code># This file contains blah, blah, blah...
# nrow 25012
# ncol 91</code></pre>
<p>The number of rows (<code>nrow</code>) includes <em>only</em> the data rows (<em>not</em> the comment rows, <em>nor</em> the row with variable names). On the other hand, the number of columns (<code>ncol</code>) <em>does include</em> the column with individual IDs.</p>
<p>All of these CSV files may be transposed relative to the form described below. You just need to include, in the control file, a line like</p>
<pre><code>geno_transposed: true</code></pre>
<div id="genotype-and-founder-genotype-data" class="section level4">
<h4>Genotype and founder genotype data</h4>
<p>The genotype data file is a matrix of individuals × markers. The first column is the individual IDs; the first row is the marker names. The founder genotypes (if needed) are in the same form, with founder lines as rows and markers as columns, and with founder IDs in the first column.</p>
</div>
<div id="phenotype-and-covariate-data" class="section level4">
<h4>Phenotype and covariate data</h4>
<p>We split the numeric phenotypes from the mixed-mode covariates, as two separate CSV files. Each file forms a matrix of individuals × phenotypes (or covariates), with the first column being individual IDs and the first row being phenotype or covariate names. Sex and line IDs (if needed) can be columns in the covariate data.</p>
</div>
<div id="phenotype-covariates" class="section level4">
<h4>Phenotype covariates</h4>
<p>A separate CSV file contains phenotype covariate data, as phenotypes × phenotype covariates. The first column contains phenotype names, and the first row contains the names of the phenotype covariates.</p>
<p><strong>Note</strong>: The genotype, founder genotype, phenotype, covariate, and phenotype covariate data can be split across multiple files. For example, the genotype data could be split by chromosome. The individual IDs must appear in each file; these are used to combine the files.</p>
</div>
<div id="genetic-and-physical-maps" class="section level4">
<h4>Genetic and physical maps</h4>
<p>Genetic and physical maps of the genotyped markers will be as separate CSV files, each with three columns: marker, chromosome, and position. The first row should be <code>marker,chr,pos</code> but will be ignored. In the genetic map file, positions should be in centiMorgans (cM). In the physical map file, positions should be in megabasepairs (Mbp).</p>
</div>
<div id="cross-information" class="section level4">
<h4>Cross information</h4>
<p>The <code>&quot;cross_info&quot;</code> data specifies details of the cross that generated each individual and is a numeric matrix with individuals as rows (the same number of rows as in the genotype data) and with columns depending on the cross type.</p>
<p>For simple cross types (e.g., <code>&quot;f2&quot;</code>, an intercross between two inbred lines), this cross information may be included as a column in the covariate data. More generally, the cross information will be a separate CSV file. For example, for a set of Collaborative Cross (CC) lines, we will want a matrix with eight columns, which indicate the order of the founders in the crosses that generated each CC line.</p>
<p>So, in general, the cross information will be in a CSV file with individuals as rows and a set of columns that define the cross information for that cross type. The first column contains individual IDs and the first row contains column names. Details on the column information are provided in the cross-type-specific information, below.</p>
</div>
</div>
<div id="control-file" class="section level3">
<h3>Control file</h3>
<p>The new input file format includes a text-based control file (in <a href="http://www.yaml.org">YAML</a> or <a href="http://www.json.org">JSON</a> format) to specify the names of all of the other files as well as various control parameters such as genotype and sex encodings and codes for missing values. We use <a href="http://www.yaml.org">YAML</a> because it is flexible, readable, and easy to import into R. We also allow <a href="http://www.json.org">JSON</a>; though it is often less human-readable, it can be less prone to errors.</p>
<p>The format of the control file is a bit technical. We describe the details here and also provide a function <code>write_control_file()</code> that takes the detailed specifications as input and contructs the control file in the correct format.</p>
<p>We’ll start with an example: the control file for the <a href="http://kbroman.org/qtl2/pages/sampledata">sample intercross data</a>.</p>
<pre><code># Data from Grant et al. (2006) Hepatology 44:174-185
# Abstract of paper at PubMed: https://www.ncbi.nlm.nih.gov/pubmed/16799992
# Available as part of R/qtl book package, https://github.com/kbroman/qtlbook
crosstype: f2
geno: iron_geno.csv
pheno: iron_pheno.csv
phenocovar: iron_phenocovar.csv
covar: iron_covar.csv
gmap: iron_gmap.csv
alleles:
- S
- B
genotypes:
  SS: 1
  SB: 2
  BB: 3
sex:
  covar: sex
  f: female
  m: male
cross_info:
  covar: cross_direction
  (SxB)x(SxB): 0
  (BxS)x(BxS): 1
x_chr: X
na.strings:
- '-'
- NA</code></pre>
<p>Any line that begins with a “<code>#</code>” is treated as a comment and ignored. It’s good to include some comments at the top of the file, describing the dataset.</p>
<p>The order of things within the file is not important, but the names of things are critical.</p>
<p>Much of the information is represented as key-value pairs, as “<code>key: value</code>.” For example, the cross type is indicated with a line like</p>
<pre><code>crosstype: f2</code></pre>
<p>The “<code>key</code>” is “<code>crosstype</code>” and the “<code>value</code>” is “<code>f2</code>.” This indicates that the data are for an F<sub>2</sub> intercross between two inbred lines.</p>
<div id="file-names" class="section level4">
<h4>File names</h4>
<p>The names of the basic CSV files are indicated with lines like</p>
<pre><code>geno: iron_geno.csv</code></pre>
<p>This indicates that the genotype data are in the file <code>iron_geno.csv</code>. The files are expected to be in the same directory as the control file. They could be placed in separate directories, with the file names being paths relative to the location of the control file, but this is not recommended (or well tested).</p>
<p>The “keys” for the different files are the following:</p>
<pre><code>geno:         genotype_filename
founder_geno: founder_genotype_filename
pheno:        phenotype_filename
covar:        covariate_filename
phenocovar:   phenotype_covariate_filename
gmap:         genetic_map_filename
pmap:         physical_map_filename</code></pre>
<p>Most of these files are optional; if a particular file is not used, the corresponding key can be omitted from the control file.</p>
<p>If the data for a section is split into multiple files (for example, if the genotypes are split into chromosome-specific files), then a vector of file names should be provided. For example:</p>
<pre><code>geno:
  - geno1.csv
  - geno2.csv
  - genoX.csv
founder_geno:
  - founder_geno1.csv
  - founder_geno2.csv
  - founder_genoX.csv</code></pre>
</div>
<div id="x-chromosome" class="section level4">
<h4>X chromosome</h4>
<p>If one of the chromosomes is to be treated as the X chromosome, there should be a line like</p>
<pre><code>x_chr: X</code></pre>
<p>This specifies the chromosome ID for the X chromosome (<code>X</code> in this case).</p>
</div>
<div id="allele-labels" class="section level4">
<h4>Allele labels</h4>
<p>To add labels in summary tables and plots, provide a vector of single-character allele labels, with one for each founder line. For example,</p>
<pre><code>alleles:
- S
- B</code></pre>
<p>This list of items, each beginning with a hyphen and a space, is the YAML format for a vector. It is equivalent to the R code <code>c(&quot;S&quot;, &quot;B&quot;)</code>.</p>
<p>You could also write this line as</p>
<pre><code>alleles: [S, B]</code></pre>
<p>which is an alternative format for vectors in YAML.</p>
</div>
<div id="genotype-codes" class="section level4">
<h4>Genotype codes</h4>
<p>The control file should contain a record with “<code>genotypes:</code>” that specifies the genotype encodings. Here’s an example:</p>
<pre><code>genotypes:
  SS: 1
  SB: 2
  BB: 3</code></pre>
<p>For each possible genotype code, indent and provide a “<code>key: value</code>” pair, with the key being the code used in the genotype and founder genotype files, and the value being an integer to which the genotype should be converted.</p>
<p>The above example would be suitable for a backcross or intercross. For a backcross, the second homozygote (<code>BB</code> in this case) is only needed in the case that there are X chromosome genotypes for males.</p>
<p>For RIL, we would use something like</p>
<pre><code>genotypes:
  BB: 1
  DD: 2</code></pre>
<p>For crosses with multiple parents, the genotype file should contain genotype calls for a set of SNPs, and there should be a corresponding founder genotype file with genotypes of the founders at those SNPs. A common set of genotype codes needs to be used for all SNPs. In particular, the genotypes <strong>cannot</strong> be encoded as <code>AA</code>, <code>CC</code>, <code>GG</code>, <code>TT</code>, <code>AC</code>, <code>AG</code>, because then, e.g., <code>CC</code> would need to be treated as <code>1</code> for some SNPs and <code>3</code> for others. Instead, code the genotypes with something like <code>AA</code>, <code>AB</code>, <code>BB</code>, and then include the following in the control file:</p>
<pre><code>genotypes:
  AA: 1
  AB: 2
  BB: 3</code></pre>
</div>
<div id="sex" class="section level4">
<h4>Sex</h4>
<p>Sex can be provided as a column in the covariate file or as a separate file.</p>
<p>If it is a column in the covariate file, the control file should have a section that looks like this:</p>
<pre><code>sex:
  covar: sex
  f: female
  m: male</code></pre>
<p>Here, “<code>covar: sex</code>” indicates that the column name used in the covariate file is “<code>sex</code>.” If the column name were “<code>Sex</code>,” you would write “<code>covar: Sex</code>.”</p>
<p>The other two “<code>key: value</code>” pairs are the encodings used for sex, with the “keys” being the codes used in the covariate file and the “values” being <code>female</code> and <code>male</code>. So this indicates that sex was encoded as <code>f</code> for females and <code>m</code> for males. If, instead, the sex covariate had <code>0</code> for females and <code>1</code> for males, you would use:</p>
<pre><code>sex:
  covar: sex
  0: female
  1: male</code></pre>
<p>Sex information can also be provided as a separate file. In this case, the file should have two columns: individual ID, and sex. Further, the part of the control file dealing with sex should look like this:</p>
<pre><code>sex:
  file: sex_filename
  f: female
  m: male</code></pre>
<p>So instead of a line with “<code>covar:</code>,” use “<code>file:</code>” followed by the name of the file (e.g., “<code>file: iron_sex.csv</code>”). You must still provide the sex encodings, as before.</p>
</div>
<div id="cross-information-1" class="section level4">
<h4>Cross information</h4>
<p>For simple crosses (e.g., an intercross), cross information can be a single column within the covariate file. In this case, include something like the following in the control file:</p>
<pre><code>cross_info:
  covar: cross_direction
  (SxB)x(SxB): 0
  (BxS)x(BxS): 1</code></pre>
<p>This is much like the information for sex. The “<code>covar:</code>” line indicates the name of the column in the covariate data that corresponds to the cross information. The other two lines indicate the encodings of the cross information as “<code>key: value</code>” pairs, where “<code>key</code>” is the code used in the cross information column and “<code>value</code>” is the integer to which it should be converted.</p>
<p>More generally, the cross information would be contained in a separate comma-delimited file. For simple crosses, in which the cross information is a single column, we allow it to be encoded differently from what is needed, and the control file information should look like this:</p>
<pre><code>cross_info:
  file: crossinfo_filename
  (SxB)x(SxB): 0
  (BxS)x(BxS): 1</code></pre>
<p>For more complex crosses (e.g., the Collaborative Cross), the cross information spans multiple columns and we require that the user have set this up in advance (i.e., no translation of encodings will be performed). In this case the relevant section of the control file looks like this:</p>
<pre><code>cross_info:
  file: crossinfo_filename</code></pre>
<p>Or, more simply, you could write:</p>
<pre><code>cross_info: crossinfo_filename</code></pre>
</div>
<div id="missing-value-codes" class="section level4">
<h4>Missing value codes</h4>
<p>To indicate the set of codes that are to be treated as missing values in the genotype, founder genotype, phenotype, covariate, and phenotype covariate files, define <code>na.strings</code> within the control file:</p>
<pre><code>na.strings:
- NA
- '-'</code></pre>
<p>A hyphen needs to be surrounded in single- or double-quotes. Many other character strings (such as <code>NA</code>) do not. This is a similar contruction as for the allele codes above; the list with hyphens followed a space is the YAML format for a vector. You could also write:</p>
<pre><code>na.strings: [NA, '-']</code></pre>
<p>which is another way to define a vector with YAML.</p>
</div>
<div id="field-separator" class="section level4">
<h4>Field separator</h4>
<p>If the data files use a separator other than a comma (e.g., a semi-colon, or the vertical bar (<code>|</code>) which I like because it is seldom present in data), indicate the separator within the control file, as follows:</p>
<pre><code>sep: '|'</code></pre>
<p>A vertical bar needs to be surrounded by single- or double-quotes. A semicolon doesn’t, but it doesn’t hurt if you do.</p>
</div>
</div>
</div>
<div id="detailed-specifications-for-each-cross-type" class="section level2">
<h2>Detailed specifications for each cross type</h2>
<p>[to be provided, eventually]</p>
</div>



<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
