<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">

<head>

<meta charset="utf-8" />
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="generator" content="pandoc" />

<meta name="viewport" content="width=device-width, initial-scale=1">



<title>Differences between R/qtl and R/qtl2</title>



<style type="text/css">code{white-space: pre;}</style>
<style type="text/css">
div.sourceCode { overflow-x: auto; }
table.sourceCode, tr.sourceCode, td.lineNumbers, td.sourceCode {
  margin: 0; padding: 0; vertical-align: baseline; border: none; }
table.sourceCode { width: 100%; line-height: 100%; }
td.lineNumbers { text-align: right; padding-right: 4px; padding-left: 4px; color: #aaaaaa; border-right: 1px solid #aaaaaa; }
td.sourceCode { padding-left: 5px; }
code > span.kw { color: #007020; font-weight: bold; } /* Keyword */
code > span.dt { color: #902000; } /* DataType */
code > span.dv { color: #40a070; } /* DecVal */
code > span.bn { color: #40a070; } /* BaseN */
code > span.fl { color: #40a070; } /* Float */
code > span.ch { color: #4070a0; } /* Char */
code > span.st { color: #4070a0; } /* String */
code > span.co { color: #60a0b0; font-style: italic; } /* Comment */
code > span.ot { color: #007020; } /* Other */
code > span.al { color: #ff0000; font-weight: bold; } /* Alert */
code > span.fu { color: #06287e; } /* Function */
code > span.er { color: #ff0000; font-weight: bold; } /* Error */
code > span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
code > span.cn { color: #880000; } /* Constant */
code > span.sc { color: #4070a0; } /* SpecialChar */
code > span.vs { color: #4070a0; } /* VerbatimString */
code > span.ss { color: #bb6688; } /* SpecialString */
code > span.im { } /* Import */
code > span.va { color: #19177c; } /* Variable */
code > span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code > span.op { color: #666666; } /* Operator */
code > span.bu { } /* BuiltIn */
code > span.ex { } /* Extension */
code > span.pp { color: #bc7a00; } /* Preprocessor */
code > span.at { color: #7d9029; } /* Attribute */
code > span.do { color: #ba2121; font-style: italic; } /* Documentation */
code > span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code > span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code > span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
</style>



<link href="data:text/css;charset=utf-8,body%20%7B%0Abackground%2Dcolor%3A%20%23fff%3B%0Amargin%3A%201em%20auto%3B%0Amax%2Dwidth%3A%20700px%3B%0Aoverflow%3A%20visible%3B%0Apadding%2Dleft%3A%202em%3B%0Apadding%2Dright%3A%202em%3B%0Afont%2Dfamily%3A%20%22Open%20Sans%22%2C%20%22Helvetica%20Neue%22%2C%20Helvetica%2C%20Arial%2C%20sans%2Dserif%3B%0Afont%2Dsize%3A%2014px%3B%0Aline%2Dheight%3A%201%2E35%3B%0A%7D%0A%23header%20%7B%0Atext%2Dalign%3A%20center%3B%0A%7D%0A%23TOC%20%7B%0Aclear%3A%20both%3B%0Amargin%3A%200%200%2010px%2010px%3B%0Apadding%3A%204px%3B%0Awidth%3A%20400px%3B%0Aborder%3A%201px%20solid%20%23CCCCCC%3B%0Aborder%2Dradius%3A%205px%3B%0Abackground%2Dcolor%3A%20%23f6f6f6%3B%0Afont%2Dsize%3A%2013px%3B%0Aline%2Dheight%3A%201%2E3%3B%0A%7D%0A%23TOC%20%2Etoctitle%20%7B%0Afont%2Dweight%3A%20bold%3B%0Afont%2Dsize%3A%2015px%3B%0Amargin%2Dleft%3A%205px%3B%0A%7D%0A%23TOC%20ul%20%7B%0Apadding%2Dleft%3A%2040px%3B%0Amargin%2Dleft%3A%20%2D1%2E5em%3B%0Amargin%2Dtop%3A%205px%3B%0Amargin%2Dbottom%3A%205px%3B%0A%7D%0A%23TOC%20ul%20ul%20%7B%0Amargin%2Dleft%3A%20%2D2em%3B%0A%7D%0A%23TOC%20li%20%7B%0Aline%2Dheight%3A%2016px%3B%0A%7D%0Atable%20%7B%0Amargin%3A%201em%20auto%3B%0Aborder%2Dwidth%3A%201px%3B%0Aborder%2Dcolor%3A%20%23DDDDDD%3B%0Aborder%2Dstyle%3A%20outset%3B%0Aborder%2Dcollapse%3A%20collapse%3B%0A%7D%0Atable%20th%20%7B%0Aborder%2Dwidth%3A%202px%3B%0Apadding%3A%205px%3B%0Aborder%2Dstyle%3A%20inset%3B%0A%7D%0Atable%20td%20%7B%0Aborder%2Dwidth%3A%201px%3B%0Aborder%2Dstyle%3A%20inset%3B%0Aline%2Dheight%3A%2018px%3B%0Apadding%3A%205px%205px%3B%0A%7D%0Atable%2C%20table%20th%2C%20table%20td%20%7B%0Aborder%2Dleft%2Dstyle%3A%20none%3B%0Aborder%2Dright%2Dstyle%3A%20none%3B%0A%7D%0Atable%20thead%2C%20table%20tr%2Eeven%20%7B%0Abackground%2Dcolor%3A%20%23f7f7f7%3B%0A%7D%0Ap%20%7B%0Amargin%3A%200%2E5em%200%3B%0A%7D%0Ablockquote%20%7B%0Abackground%2Dcolor%3A%20%23f6f6f6%3B%0Apadding%3A%200%2E25em%200%2E75em%3B%0A%7D%0Ahr%20%7B%0Aborder%2Dstyle%3A%20solid%3B%0Aborder%3A%20none%3B%0Aborder%2Dtop%3A%201px%20solid%20%23777%3B%0Amargin%3A%2028px%200%3B%0A%7D%0Adl%20%7B%0Amargin%2Dleft%3A%200%3B%0A%7D%0Adl%20dd%20%7B%0Amargin%2Dbottom%3A%2013px%3B%0Amargin%2Dleft%3A%2013px%3B%0A%7D%0Adl%20dt%20%7B%0Afont%2Dweight%3A%20bold%3B%0A%7D%0Aul%20%7B%0Amargin%2Dtop%3A%200%3B%0A%7D%0Aul%20li%20%7B%0Alist%2Dstyle%3A%20circle%20outside%3B%0A%7D%0Aul%20ul%20%7B%0Amargin%2Dbottom%3A%200%3B%0A%7D%0Apre%2C%20code%20%7B%0Abackground%2Dcolor%3A%20%23f7f7f7%3B%0Aborder%2Dradius%3A%203px%3B%0Acolor%3A%20%23333%3B%0Awhite%2Dspace%3A%20pre%2Dwrap%3B%20%0A%7D%0Apre%20%7B%0Aborder%2Dradius%3A%203px%3B%0Amargin%3A%205px%200px%2010px%200px%3B%0Apadding%3A%2010px%3B%0A%7D%0Apre%3Anot%28%5Bclass%5D%29%20%7B%0Abackground%2Dcolor%3A%20%23f7f7f7%3B%0A%7D%0Acode%20%7B%0Afont%2Dfamily%3A%20Consolas%2C%20Monaco%2C%20%27Courier%20New%27%2C%20monospace%3B%0Afont%2Dsize%3A%2085%25%3B%0A%7D%0Ap%20%3E%20code%2C%20li%20%3E%20code%20%7B%0Apadding%3A%202px%200px%3B%0A%7D%0Adiv%2Efigure%20%7B%0Atext%2Dalign%3A%20center%3B%0A%7D%0Aimg%20%7B%0Abackground%2Dcolor%3A%20%23FFFFFF%3B%0Apadding%3A%202px%3B%0Aborder%3A%201px%20solid%20%23DDDDDD%3B%0Aborder%2Dradius%3A%203px%3B%0Aborder%3A%201px%20solid%20%23CCCCCC%3B%0Amargin%3A%200%205px%3B%0A%7D%0Ah1%20%7B%0Amargin%2Dtop%3A%200%3B%0Afont%2Dsize%3A%2035px%3B%0Aline%2Dheight%3A%2040px%3B%0A%7D%0Ah2%20%7B%0Aborder%2Dbottom%3A%204px%20solid%20%23f7f7f7%3B%0Apadding%2Dtop%3A%2010px%3B%0Apadding%2Dbottom%3A%202px%3B%0Afont%2Dsize%3A%20145%25%3B%0A%7D%0Ah3%20%7B%0Aborder%2Dbottom%3A%202px%20solid%20%23f7f7f7%3B%0Apadding%2Dtop%3A%2010px%3B%0Afont%2Dsize%3A%20120%25%3B%0A%7D%0Ah4%20%7B%0Aborder%2Dbottom%3A%201px%20solid%20%23f7f7f7%3B%0Amargin%2Dleft%3A%208px%3B%0Afont%2Dsize%3A%20105%25%3B%0A%7D%0Ah5%2C%20h6%20%7B%0Aborder%2Dbottom%3A%201px%20solid%20%23ccc%3B%0Afont%2Dsize%3A%20105%25%3B%0A%7D%0Aa%20%7B%0Acolor%3A%20%230033dd%3B%0Atext%2Ddecoration%3A%20none%3B%0A%7D%0Aa%3Ahover%20%7B%0Acolor%3A%20%236666ff%3B%20%7D%0Aa%3Avisited%20%7B%0Acolor%3A%20%23800080%3B%20%7D%0Aa%3Avisited%3Ahover%20%7B%0Acolor%3A%20%23BB00BB%3B%20%7D%0Aa%5Bhref%5E%3D%22http%3A%22%5D%20%7B%0Atext%2Ddecoration%3A%20underline%3B%20%7D%0Aa%5Bhref%5E%3D%22https%3A%22%5D%20%7B%0Atext%2Ddecoration%3A%20underline%3B%20%7D%0A%0Acode%20%3E%20span%2Ekw%20%7B%20color%3A%20%23555%3B%20font%2Dweight%3A%20bold%3B%20%7D%20%0Acode%20%3E%20span%2Edt%20%7B%20color%3A%20%23902000%3B%20%7D%20%0Acode%20%3E%20span%2Edv%20%7B%20color%3A%20%2340a070%3B%20%7D%20%0Acode%20%3E%20span%2Ebn%20%7B%20color%3A%20%23d14%3B%20%7D%20%0Acode%20%3E%20span%2Efl%20%7B%20color%3A%20%23d14%3B%20%7D%20%0Acode%20%3E%20span%2Ech%20%7B%20color%3A%20%23d14%3B%20%7D%20%0Acode%20%3E%20span%2Est%20%7B%20color%3A%20%23d14%3B%20%7D%20%0Acode%20%3E%20span%2Eco%20%7B%20color%3A%20%23888888%3B%20font%2Dstyle%3A%20italic%3B%20%7D%20%0Acode%20%3E%20span%2Eot%20%7B%20color%3A%20%23007020%3B%20%7D%20%0Acode%20%3E%20span%2Eal%20%7B%20color%3A%20%23ff0000%3B%20font%2Dweight%3A%20bold%3B%20%7D%20%0Acode%20%3E%20span%2Efu%20%7B%20color%3A%20%23900%3B%20font%2Dweight%3A%20bold%3B%20%7D%20%20code%20%3E%20span%2Eer%20%7B%20color%3A%20%23a61717%3B%20background%2Dcolor%3A%20%23e3d2d2%3B%20%7D%20%0A" rel="stylesheet" type="text/css" />

</head>

<body>




<h1 class="title toc-ignore">Differences between R/qtl and R/qtl2</h1>



<p><a href="http://kbroman.org/qtl2">R/qtl2</a> is a reimplementation of the QTL analysis software <a href="https://rqtl.org">R/qtl</a>, to better handle high-dimensional data and complex cross designs.</p>
<p>There have been a number of big changes. We summarize some of them here in order to assist long-time R/qtl users in their transition to R/qtl2.</p>
<div id="new-data-file-formats" class="section level2">
<h2>New data file formats</h2>
<p>The input file format has been completely changed. R/qtl2 allows just a single format, with data split across multiple CSV files, and with a control file (in <a href="http://www.yaml.org">YAML</a> or <a href="http://www.json.org">JSON</a> format).</p>
<p>For more information about the new file format, see the <a href="http://kbroman.org/qtl2/assets/vignettes/input_files.html">related vignette</a>, as well as <a href="http://kbroman.org/qtl2/pages/sampledata.html">sample input data files</a> and the <a href="https://github.com/rqtl/qtl2data">qtl2data repository</a>.</p>
<p>R/qtl cross objects can be converted to the new format with the <code>qtl2geno::convert2cross2()</code> function, and so one can continue to use the R/qtl function <code>read.cross()</code> to read in data and then convert them to the new format:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(qtl)
mycross &lt;-<span class="st"> </span><span class="kw">read.cross</span>(<span class="st">&quot;csv&quot;</span>, <span class="dt">file=</span><span class="st">&quot;mydata.csv&quot;</span>)
<span class="kw">library</span>(qtl2geno)
mycross &lt;-<span class="st"> </span><span class="kw">convert2cross2</span>(mycross)</code></pre></div>
</div>
<div id="new-data-structures" class="section level2">
<h2>New data structures</h2>
<p>The data structures used by R/qtl2 are completely different than those used by R/qtl. The new data structures are no simpler than before, but they tend to be a bit “flatter” (that is, less deeply nested). Most everything is a “list”, and we’re using fewer “attributes” and instead including such things as components at the top level.</p>
</div>
<div id="split-into-multiple-packages" class="section level2">
<h2>Split into multiple packages</h2>
<p>R/qtl2 is not a single package as R/qtl is. Rather it’s split into multiple packages:</p>
<ul>
<li><a href="https://github.com/rqtl/qtl2geno">qtl2geno</a>, for calculating genotype probabilities, imputations, and genetic maps</li>
<li><a href="https://github.com/rqtl/qtl2scan">qtl2scan</a>, for QTL genome scans and related calculations</li>
<li><a href="https://github.com/rqtl/qtl2plot">qtl2plot</a>, for data visualization</li>
<li><a href="https://github.com/rqtl/qtl2convert">qtl2convert</a>, for converting data among the R/qtl2, R/qtl, and <a href="https://www.bioconductor.org/packages/release/bioc/html/DOQTL.html">DOQTL</a> formats</li>
</ul>
<p>We have in mind that, for high-dimensional data, the QTL genotype probability calculations with qtl2geno will be performed once and saved, and that the genome scans with qtl2scan will be performed in “batch” (e.g., on a cluster) and also saved, and that interactive analyses will mostly be in the data visualizations with qtl2plot.</p>
<p>It can be confusing to remember which function is in which package. For this reason, we created an additional, largely empty package <a href="https://github.com/rqtl/qtl2">qtl2</a>. If you load qtl2 with <code>library(qtl2)</code>, the three main packages, <a href="https://github.com/rqtl/qtl2geno">qtl2geno</a>, <a href="https://github.com/rqtl/qtl2scan">qtl2scan</a>, and <a href="https://github.com/rqtl/qtl2plot">qtl2plot</a>, will all be loaded.</p>
</div>
<div id="new-functions-names" class="section level2">
<h2>New functions names</h2>
<p>The names of all of the main functions have changed, mostly with periods replaced by underscores. For example:</p>
<ul>
<li><code>read.cross()</code> → <code>read_cross2()</code></li>
<li><code>calc.genoprob()</code> → <code>calc_genoprob()</code></li>
<li><code>sim.geno()</code> → <code>sim_geno()</code></li>
<li><code>argmax.geno()</code> → <code>viterbi()</code></li>
<li><code>scanone()</code> → <code>scan1()</code></li>
<li><code>nmar()</code> → <code>n_mar()</code></li>
<li><code>phenames()</code> → <code>pheno_names()</code></li>
<li><code>markernames()</code> → <code>marker_names()</code></li>
</ul>
</div>
<div id="treatment-of-intermediate-calculations" class="section level2">
<h2>Treatment of intermediate calculations</h2>
<p>We’re no longer storing intermediate calculations as part of the cross object. For example, <code>calc_genoprob()</code>, to calculate QTL genotype probabilities given the observed marker data, returns a list with the probabilities. <code>scan1()</code>, to perform a genome scan, takes these probabilities plus a phenotype matrix.</p>
</div>
<div id="use-of-individual-identifiers-for-aligning-data" class="section level2">
<h2>Use of individual identifiers for aligning data</h2>
<p>Individual identifiers are now used to ensure the alignment of individuals, for example between the QTL genotype probabilties and the phenotype data.</p>
<p>For example, in <code>scan1()</code>, to perform a genome scan, it’s necessary that the phenotype data carry the corresponding individual IDs as row names.</p>
<p>As a result, when subsetting out, say, females, when calling <code>scan1</code>, you only need to subset one of the inputs, and the rest will be automatically subset for you.</p>
<pre><code>out_all &lt;- scan1(probs, pheno, kinship)
out_f &lt;- scan1(probs, pheno[sex==&quot;female&quot;,], kinship)</code></pre>
</div>
<div id="order-of-arguments-when-subsetting-cross-objects" class="section level2">
<h2>Order of arguments when subsetting cross objects</h2>
<p>In R/qtl, when subsetting a cross object, you can use square brackets, like this:</p>
<pre><code>library(qtl)
mycross[chr, ind]</code></pre>
<p>But the order of those two arguments was not very well chosen. It’s better to think of individuals as rows and chromosomes as column, and so put the individuals first.</p>
<p>And so in R/qtl2, we’ve switched the order of arguments: in bracket subsetting of cross objects, individuals now come first.</p>
<pre><code>library(qtl2geno)
mycross[ind, chr]</code></pre>
</div>
<div id="in-scan-of-x-chr-need-to-provide-special-covariates" class="section level2">
<h2>In scan of X chr, need to provide special covariates</h2>
<p>In R/qtl, when you perform a single-QTL scan of the X chromosome, it identifies appropriate covariates to include, to avoid spurious linkage due to sex and cross-direction differences.</p>
<p>In R/qtl2, you need to provide such covariates yourself via the <code>Xcovar</code> argument to <code>scan1()</code>, There is a function in qtl2geno, <code>get_x_covar</code>, for deriving these, but you’re a bit more on your own.</p>
</div>



<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
